### PR TITLE
[FLINK-11918][table] Deprecated Window and Rename it to GroupWindow

### PR DESCRIPTION
**Pull request by hequn8128**
## What is the purpose of the change

This pull request deprecates `Window` and Rename it to `GroupWindow`.

## Brief change log

  - Deprecate table.window(window: Window) and add table.window(window: GroupWindow)
  - Deprecate Window class and Add GroupWindow class.
  - Deprecate WindowedTable class and Add GroupWindowedTable class.

## Verifying this change

This change is already covered by existing tests, such as `GroupWindowITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no, but contains api change)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by twalthr** [comment]
@flinkbot approve-until quality

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/table.scala [hunk 1]
```diff
@@ -1255,6 +1288,14 @@ class WindowedTable(
   }
 }
 
+/**
+  * A table that has been windowed for grouping [[Window]]s.
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/table.scala]: Replace with `[[GroupWindow]]`

**Comment by sunjincheng121** [comment]
> Thank you @hequn8128.
> 
> What about `Tumble`, `Session`, `Slide`, and `Over` window? Do we also want to deprecate `org.apache.flink.table.api.java.Tumble` and encourage `org.apache.flink.table.api.Tumble`?

IMHO. It is important to correct the concept of `Window` to `GroupWindow`. The package changes for these auxiliary methods can be kept in 1.9. Users seeing many classes with the same name will be very confused，similar to current `TableEnvironment`s.  So, I suggest removing them directly in FLINK-11908.

What do you think? @twalthr  @hequn8128

**Comment by twalthr** [comment]
I think the overhead of performing this change in 1.8 as well is pretty low. I agree that this might be confusing for users in 1.8 but this must be done at some point and the next release will make the situation better as there is only one class for each window left.

**Comment by sunjincheng121** [comment]
I meant we can using one class in Flink master directly. I think we can accept a package break.

**Comment by twalthr** [comment]
If we can accept a package break, we could also accept the break of `Window` to `GroupWindow` because this class is also not used by a user directly.

**Comment by sunjincheng121** [comment]
Ah, Maybe you are right, but I think`Window` and `GroupWindow` affects the interface of the Table API. `def window(window: Window): WindowedTable`. which is a big change for users.

**Comment by twalthr** [comment]
`Tumble`, `Session` etc. are also part of the Table API.

The change performed in this PR is not a big change for users as programs such as `.window(Tumble over...)` are still source code compatible and compile. However, if we move `Tumble`, users will need to update their imports which is a big change.

**Comment by sunjincheng121** [comment]
Ooh, @twalthr, I got your points! I think we are on the same page about put the expiration method to 1.8 as soon as possible.

I mainly concern is many classes with the same name will be very confused.  And in this PR we deal with the concept of `Window`.  if we want to deal with the `Tumble`, `Session` etc in this PR.  and merge into release-1.8,  then let API fully compatible, that's the correct thing we should follow.  I agree with your proposal. 

+1 from the points of my view.

Best,
Jincheng

**Comment by sunjincheng121** [comment]
Hi @hequn8128 is that make sense to you?  I appreciate if you can update the PR according to @twalthr 's proposal.

Thanks,
Jincheng

**Comment by hequn8128** [comment]
@sunjincheng121 @twalthr Thanks for your suggestions. Putting the deprecated APIs to 1.8 as soon as possible is a good idea. I will update the PR asap. 

Thanks, Hequn

**Comment by hequn8128** [comment]
@twalthr @sunjincheng121 I have addressed all your comments and updated the PR. Would be nice if you can take another look. Thank you.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/WindowHelper.scala [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+package org.apache.flink.table.api
+
+import org.apache.flink.table.expressions.{Expression, ExpressionParser}
+
+/**
+  * Base class for Tumble Window Helper classes. This class contains help methods to create Tumble
+  * Windows.
+  */
+trait TumbleBase {
```
Comment (sunjincheng121) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/WindowHelper.scala]: I think we do not need create a new class `WindowHelper` all those defined can be placed into `winodws.scala` What do you think?

Comment (hequn8128) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/WindowHelper.scala]: Make sense!

**Comment by hequn8128** [comment]
@twalthr Thank you a lot for the review. I will address the comments from @sunjincheng121 right now :-)

**Comment by sunjincheng121** [comment]
Thanks to @hequn8128 @twalthr!
Merging ...  until CI turns green.

**Comment by twalthr** [comment]
@sunjincheng121 when merging don't forget to add release notes to JIRA for this API change.

**Comment by sunjincheng121** [comment]
Thanks for the reminder @twalthr!

**Pull request closed without merging**